File: src/utils/nexus-nl-parser.ts
/**
 * Nexus Main Mode Natural Language Parser
 *
 * Maps plain English sentences to structured Nexus SDLC intents.
 * Rule-based pattern matching with optional LLM fallback.
 */

import { env, hasAzureOpenAiConfig } from "../config/env";

export interface NexusIntent {
  command: string;
  args: string[];
  raw: string;
  source: "rule" | "llm" | "unknown";
}

interface PatternRule {
  patterns: RegExp[];
  extract: (match: RegExpMatchArray, raw: string) => NexusIntent;
}

const RULES: PatternRule[] = [
  // Tickets
  {
    patterns: [
      /^tickets$/i,
      /^(?:list|show|get|what\s+are)\s+(?:the\s+)?tickets\??$/i,
    ],
    extract: (_m, raw) => ({ command: "tickets", args: [], raw, source: "rule" }),
  },
  // Plan
  {
    patterns: [
      /^(?:generate|create|make|build|get)\s+(?:a\s+)?(?:more\s+)?(?:comprehensive|detailed|detail)\s+plan\s+for\s+(\S+)$/i,
      /^plan\s+(\S+)$/i,
      /^(?:create|make|build|generate)\s+(?:a\s+)?plan\s+for\s+(\S+)$/i,
      /^what\s+is\s+the\s+plan\s+for\s+(\S+)\??$/i,
    ],
    extract: (m, raw) => {
      const ticketId = m[1];
      const isDetailed = raw.toLowerCase().includes("comprehensive") || raw.toLowerCase().includes("detailed") || raw.toLowerCase().includes("detail");
      return {
        command: "plan",
        args: [ticketId, isDetailed ? "detailed" : "basic"],
        raw,
        source: "rule",
      };
    },
  },
  // Execute
  {
    patterns: [
      /^execute\s+(\S+)$/i,
      /^(?:run|start|execute|process)\s+(\S+)$/i,
      /^(?:work\s+on|implement)\s+(\S+)$/i,
    ],
    extract: (m, raw) => ({ command: "execute", args: [m[1]], raw, source: "rule" }),
  },
  // Status
  {
    patterns: [
      /^status$/i,
      /^(?:show|what\s+is)\s+(?:the\s+)?(?:current\s+)?status\??$/i,
      /^how\s+are\s+we\s+doing\??$/i,
    ],
    extract: (_m, raw) => ({ command: "status", args: [], raw, source: "rule" }),
  },
  // AI Health
  {
    patterns: [
      /^ai$/i,
      /^ai\s+health$/i,
      /^(?:check|is|how\s+is)\s+(?:the\s+)?ai(?:\s+health)?\??$/i,
    ],
    extract: (_m, raw) => ({ command: "ai", args: [], raw, source: "rule" }),
  },
  // Mode Switches
  {
    patterns: [
      /^(?:go\s+to|switch\s+to|enter|open)\s+security(?:\s+mode)?$/i,
      /^security$/i,
    ],
    extract: (_m, raw) => ({ command: "security", args: [], raw, source: "rule" }),
  },
  {
    patterns: [
      /^(?:go\s+to|switch\s+to|enter|open)\s+nlp(?:\s+mode|chat)?$/i,
      /^nlp$/i,
    ],
    extract: (_m, raw) => ({ command: "nlp", args: [], raw, source: "rule" }),
  },
  {
    patterns: [
      /^(?:go\s+to|switch\s+to|enter|open)\s+devops(?:\s+mode)?$/i,
      /^devops$/i,
    ],
    extract: (_m, raw) => ({ command: "devops", args: [], raw, source: "rule" }),
  },
  {
    patterns: [
      /^(?:go\s+to|switch\s+to|enter|open)\s+git(?:\s+mode)?$/i,
      /^git$/i,
    ],
    extract: (_m, raw) => ({ command: "git", args: [], raw, source: "rule" }),
  },
  // Push
  {
    patterns: [
      /^push\s+(\S+)$/i,
      /^(?:push|upload|sync)\s+(?:the\s+)?changes\s+for\s+(\S+)$/i,
    ],
    extract: (m, raw) => ({ command: "push", args: [m[1]], raw, source: "rule" }),
  },
  // Reset
  {
    patterns: [
      /^reset\s+(\S+)$/i,
      /^(?:reset|clear)\s+ticket\s+(\S+)$/i,
    ],
    extract: (m, raw) => ({ command: "reset", args: [m[1]], raw, source: "rule" }),
  },
  {
    patterns: [
      /^reset-all$/i,
      /^(?:reset|clear)\s+all\s+tickets$/i,
    ],
    extract: (_m, raw) => ({ command: "reset-all", args: [], raw, source: "rule" }),
  },
  // Help & Exit
  {
    patterns: [/^help$/i, /^(?:what\s+can\s+you\s+do|list\s+commands)$/i],
    extract: (_m, raw) => ({ command: "help", args: [], raw, source: "rule" }),
  },
  {
    patterns: [/^exit$/i, /^quit$/i, /^bye$/i],
    extract: (_m, raw) => ({ command: "exit", args: [], raw, source: "rule" }),
  },
];

export function parseNexusIntent(input: string): NexusIntent {
  const trimmed = input.trim();
  for (const rule of RULES) {
    for (const pattern of rule.patterns) {
      const match = trimmed.match(pattern);
      if (match) return rule.extract(match, trimmed);
    }
  }
  return { command: "unknown", args: [], raw: trimmed, source: "unknown" };
}

export async function parseNexusIntentWithLlm(input: string): Promise<NexusIntent> {
  const ruleResult = parseNexusIntent(input);
  if (ruleResult.command !== "unknown") return ruleResult;
  if (env.useMock || !hasAzureOpenAiConfig()) return ruleResult;

  try {
    const endpoint = `${env.azureEndpoint}/openai/deployments/${env.azureDeployment}/chat/completions?api-version=${env.azureApiVersion}`;
    const response = await fetch(endpoint, {
      method: "POST",
      headers: {
        "Content-Type": "application/json",
        "api-key": env.azureApiKey ?? "",
      },
      body: JSON.stringify({
        messages: [
          {
            role: "system",
            content: `You are the primary command parser for the Nexus SDLC terminal. Given a natural language sentence, return JSON: {"command": string, "args": string[]}.
Valid commands: tickets, plan, execute, status, ai, security, nlp, devops, git, push, reset, reset-all, help, exit.
Arguments:
- plan: args=["ticketId", "mode"] where mode is "detailed" or "basic". If "comprehensive" or "detailed" is mentioned, use "detailed".
- execute: args=["ticketId"]
- push: args=["ticketId"]
- reset: args=["ticketId"]
All others have args=[].
If you cannot determine the intent, return {"command": "unknown", "args": []}.`,
          },
          { role: "user", content: input },
        ],
        temperature: 0,
        response_format: { type: "json_object" },
      }),
    });

    if (!response.ok) return ruleResult;
    const data = (await response.json()) as {
      choices?: Array<{ message?: { content?: string } }>;
    };
    const content = data.choices?.[0]?.message?.content;
    if (!content) return ruleResult;
    const parsed = JSON.parse(content) as { command?: string; args?: string[] };
    if (parsed.command && parsed.command !== "unknown") {
      return { command: parsed.command, args: parsed.args ?? [], raw: input, source: "llm" };
    }
    return ruleResult;
  } catch {
    return ruleResult;
  }
}

File: src/utils/nexus_nl_parser.py
"""Nexus Main Mode NL Parser for mode switching and global system/user/config operations."""
from __future__ import annotations
import re
from dataclasses import dataclass
from ..utils.llm import parse_intent_with_llm

@dataclass
class NexusIntent:
    command: str; args: list[str]; raw: str; source: str

_RULES = [
    # Core System & Health
    (r"^(?:nexus\s+)?status$|^(?:show|what\s+is)\s+(?:the\s+)?(?:current\s+)?status\??$", lambda m, r: NexusIntent("status", [], r, "rule")),
    (r"^(?:nexus\s+)?health$", lambda m, r: NexusIntent("health", [], r, "rule")),
    (r"^(?:nexus\s+)?ai$|^ai\s+health$|^(?:check|is|how\s+is)\s+(?:the\s+)?ai", lambda m, r: NexusIntent("ai", [], r, "rule")),
    (r"^(?:nexus\s+)?doctor$", lambda m, r: NexusIntent("doctor", [], r, "rule")),
    (r"^(?:nexus\s+)?version$", lambda m, r: NexusIntent("version", [], r, "rule")),

    # Mode Switches
    (r"^(?:nexus\s+)?(?:go\s+to|switch\s+to|enter|open\s+)?security$|^sec$", lambda m, r: NexusIntent("security", [], r, "rule")),
    (r"^(?:nexus\s+)?(?:go\s+to|switch\s+to|enter|open\s+)?devops$", lambda m, r: NexusIntent("devops", [], r, "rule")),
    (r"^(?:nexus\s+)?(?:go\s+to|switch\s+to|enter|open\s+)?git$", lambda m, r: NexusIntent("git", [], r, "rule")),
    (r"^(?:nexus\s+)?(?:go\s+to|switch\s+to|enter|open\s+)?agile$|^agile$", lambda m, r: NexusIntent("agile", [], r, "rule")),

    # Sub-domain direct passthroughs
    (r"^(?:nexus\s+)?agile\s+(.*)", lambda m, r: NexusIntent("agile-direct", [m.group(1)], r, "rule")),
    (r"^(?:nexus\s+)?devops\s+(.*)", lambda m, r: NexusIntent("devops-direct", [m.group(1)], r, "rule")),
    (r"^(?:nexus\s+)?git\s+(.*)", lambda m, r: NexusIntent("git-direct", [m.group(1)], r, "rule")),
    (r"^(?:nexus\s+)?security\s+(.*)", lambda m, r: NexusIntent("security-direct", [m.group(1)], r, "rule")),

    (r"^(?:nexus\s+)?(?:project|epic|story|sprint|board|jira|docs)\b.*", lambda m, r: NexusIntent("agile-direct", [r], r, "rule")),
    (r"^(?:nexus\s+)?(?:docker|k8s|jenkins|monitor|deploy|release)\b.*", lambda m, r: NexusIntent("devops-direct", [r], r, "rule")),

    # Utilities
    (r"^(?:nexus\s+)?help$|^(?:what\s+can\s+you\s+do|list\s+commands)$", lambda m, r: NexusIntent("help", [], r, "rule")),
    (r"^(?:nexus\s+)?exit$|^quit$|^bye$", lambda m, r: NexusIntent("exit", [], r, "rule")),
]

def parse_nexus_intent(text: str) -> NexusIntent:
    t = text.strip()
    for pattern, extract in _RULES:
        m = re.match(pattern, t, re.IGNORECASE)
        if m: return extract(m, t)
    return NexusIntent("unknown", [], t, "unknown")

async def parse_nexus_intent_with_llm(text: str) -> NexusIntent:
    r = parse_nexus_intent(text)
    if r.command != "unknown": return r
    prompt = '''Nexus SDLC command parser. Return JSON: {"command": str, "args": str[]}. If the user input is a greeting, conversation, general question, or natural language request to write/edit code or perform a task in the project directory (e.g., 'hi', 'hello', 'explain...', 'how to...', 'write...', 'create...', 'refactor...'), you MUST return {"command": "unknown", "args": []}. ONLY match to these exact CLI mode commands if the user specifically requests them: status, health, ai, doctor, version, help, exit, security, devops, git, agile.'''
    parsed = await parse_intent_with_llm(text, prompt)
    if parsed: return NexusIntent(parsed["command"], parsed["args"], text, "llm")
    return r
